### PR TITLE
Avoid calling `performLayout` when only the relayout boundary is different

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
In theory a render subtree shouldn't redo its layout if the incoming constraints are the same and nothing in the tree is dirty. Whether a node is a relayout boundary (or who its relayout boundary is) isn't really "inherent" and shouldn't affect the subtree's layout.

This pull request:
- Prevents `RenderObject.layout` from aggressively marking everything dirty in a subtree until the next relayout boundary is reached (instead mark everything dirty until the closest known repaint boundary in `RenderObject.markNeedsLayout`).
- Only updates the relayout boundaries if necessary, instead of performing a full layout, for clean nodes that has the same constraints.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/test/rendering/relayout_boundary_test.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+class LayoutCountWidget extends LeafRenderObjectWidget {
+  LayoutCountWidget({ Key? key }) : super(key: key);
+
+  final RenderLayoutCount renderObject = RenderLayoutCount();
+
+  int get layoutCount => renderObject.layoutCount;
+  set layoutCount(int newValue) {
+    renderObject.layoutCount = newValue;
+  }
```
Comment (goderbauer) [inline packages/flutter/test/rendering/relayout_boundary_test.dart:59]: Do you need this setter? Maybe remove it since it is a bit doggy.

Comment (goderbauer) [reply packages/flutter/test/rendering/relayout_boundary_test.dart:59]: I'd probably also remove the getter as well and just obtain the current `layoutCount` value in the test with `tester.renderObject<RenderLayoutCount>(find.byType(RenderLayoutCount)).layoutCount`. Currently the test reads a little strange with getting the layoutCount off a widget...

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/test/rendering/relayout_boundary_test.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/rendering/relayout_boundary_test.dart]: Switched to `WidgetToRenderBoxAdapter` instead

**Pull request merged**